src/transformStateWithClones.js

```diff
@@ -3,11 +3,47 @@
 /**
  * @param {Object} state
  * @param {Object[]} actions
- *
- * @return {Object[]}
+ * @returns {Object[]}
  */
-function transformStateWithClones(state, actions) {
-  // write code here
+function transformState(state, actions) {
+  const result = [];
+  let cloneOfState = { ...state };
+
+  for (const currentAction of actions) {
+    let nextState = { ...cloneOfState };
+
+    switch (currentAction.type) {
+      case 'addProperties':
+        const extraData = currentAction.extraData;
+
+        for (const key in extraData) {
+          nextState[key] = extraData[key];
+        }
+        break;
+
+      case 'removeProperties':
+        const keysToRemove = currentAction.keysToRemove;
+
+        for (const key of keysToRemove) {
+          if (nextState.hasOwnProperty(key)) {
+            delete nextState[key];
+          }
+        }
+        break;
+
+      case 'clear':
+        nextState = {};
+        break;
+
+      default:
+        break;
+    }
+
+    result.push(nextState);
+    cloneOfState = nextState;
```

Comment:
just modify `cloneOfState` instead of `nextState` and then push it to result in the end, no need for 2 variables 😉